### PR TITLE
[site] Add interactive earlgrey block diagram to site

### PR DESCRIPTION
**Pull request by jwnrt**
## Summary

This PR adds an alternative implementation of the interactive block diagram for earlgrey to the site.

Based on the excellent block diagram from this PR: https://github.com/lowRISC/opentitan/pull/17405

## Differences to the original block diagram

This version of the diagram was made to avoid introducing Node.js and its ecosystem into OpenTitan.

* Diagram styled using CSS grids, avoiding JavaScript hydration.
* Diagram specified declaratively in HTML. Should be easy to make others.
* Stats generated into a JSON file stored on the server.
* Client-side JavaScript fetches the stats and creates the tooltips.

There's a trade-off to using this diagram, but I personally think the benefits outweigh the drawbacks:

**Benefits**:

* Avoids introducing a new technology to OpenTitan (Node.js) which requires maintenance.
* Avoids adding large dependencies (Node.js, headless chrome, and puppeteer).
* Faster build time (only `fetch_block_stats.py` has to run, and ideally only nightly).
* Design/verification stages taken from the `.hjson` files, keeping them up to date.

**Drawbacks**:

* The CSS file is fairly large, though I've tried to document it thoroughly.
* Browser requires JavaScript to be enabled to see the tooltips (rest of diagram still works).

## Screenshots

Looks mostly the same as the original:

<details>
<summary>Screenshot</summary>

image

</details>

## Unresolved

Currently no styling for:

- [x] Dual-lockstep ibex
- [ ] Crossbar crosses

Also TODO:

- [ ] Would like `util/site/fetch_block_stats.py` to run nightly rather than on build.
- [ ] The JavaScript should surface errors better when it can't fetch or parse the report.
- [ ] Remove the old diagram. At the very least, disable generation in `util/site/build-site.sh`.

### CODE REVIEW

**Comment by jwnrt** [comment]
Force pushed to change `block_diagram.css`'s license comment and satisfy the license checker.

**Comment by hcallahan-lowrisc** [comment]
If this changeset does not get merged, or does not get merged soon, let's create an issue to follow up on this comment in the original PR : https://github.com/lowRISC/opentitan/pull/17405#discussion_r1137334522.
However, if this does get some traction soon, that point should become irrelevant.

**Comment by jwnrt** [comment]
Rebased. I will have a look at these remaining TODOs sometime this week.

**Comment by jwnrt** [comment]
I've added the dual-lockstep styling, plus made the Ibex text a bit bigger like in the original diagram.

File: site/block-diagram/block_diagram.css [hunk 1]
```diff
@@ -0,0 +1,847 @@
+/* Copyright lowRISC contributors. */
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/**
+ * This is a CSS library for building block diagrams out of simple semantic HTML.
+ *
+ * We're not CSS engineers, so I've done my best not to include hacks, unstable
+ * features, or massive complexity. Remaining complexity will be documented
+ * or wrapped into variables to make intent clear.
```
Comment (HU90m) [inline site/block-diagram/block_diagram.css:10]: After reading this file, I think you can safely call yourself a CSS engineer.

File: site/block-diagram/block_diagram.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+ * parameters.
+ *
+ * @param stats Object of stats for this block with these properties:
+ *     * name
+ *     * version
+ *     * design_stage (may be null)
+ *     * verification_stage (may be null)
+ *     * total_runs (may be null)
+ *     * total_passing (may be null)
+ * @return A HTMLElement tooltip built from the given stats.
```
Comment (HU90m) [inline site/block-diagram/block_diagram.js:41]: Nit:

For JSDoc style docstrings, I don't think there should be anything before the type.

```suggestion
 * @return {HTMLElement} A tooltip element built from the given stats.
```

Also it [looks like](https://jsdoc.app/tags-returns.html) the type should be surrounded by curly braces.

This applies to [parameters](https://jsdoc.app/tags-param.html) as well.

**Review by HU90m — APPROVED** [review]
This is some of the easiest to read CSS I have encountered.

*Didn't review python script because I wrote it (so assumed it is perfect 😜)*.

File: site/block-diagram/block_diagram.css (hunk 1 above)
Comment (jwnrt) [reply site/block-diagram/block_diagram.css:10]: Oh I hope not

File: site/block-diagram/block_diagram.js (hunk 2 above)
Comment (jwnrt) [reply site/block-diagram/block_diagram.js:41]: Sure, I'll update. I wasn't actually aware of JSDoc, just writing it in a doxygen style

Comment (jwnrt) [reply site/block-diagram/block_diagram.js]: Fixed.

**Comment by jwnrt** [comment]
Rebased with `fixup`s squashed.

**Review by HU90m — APPROVED** [review]
Looks great!

Thanks for the lovely JSDoc annotations.

**Comment by hcallahan-lowrisc** [comment]
@jwnrt Could you do some cross-browser testing just for any showstoppers?

**Comment by HU90m** [comment]
The only real issue appears to be the arrows continuing beyond the block in the always on domain.

Webkit (Gnome Web):

Screenshot from 2023-03-30 11-51-35

Blink (Chromium):

Screenshot from 2023-03-30 11-51-48

File: site/block-diagram/block_diagram.css [hunk 3]
```diff
@@ -60,14 +60,21 @@
...
+
     --status1-bg: #b81d13;
     --status2-bg: #efb700;
     --status3-bg: #82be00;
     --status4-bg: #008450;
     --status1-fg: var(--text-white);
     --status2-fg: var(--text-black);
     --status3-fg: var(--text-black);
-    --status4-fg: var(--text-black);
+    --status4-fg: var(--text-white);
```
Comment (marnovandermaas) [inline site/block-diagram/block_diagram.css:78]: Nit: Is this something that can be put into an earlier commit?

File: site/block-diagram/block_diagram.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+ *     * version
+ *     * design_stage (may be null)
+ *     * verification_stage (may be null)
+ *     * total_runs (may be null)
+ *     * total_passing (may be null)
+ * @returns {HTMLElement} Tooltip element built from the given stats.
+ */
+function buildTooltip(stats) {
+    let name = stats.name;
+    // Add version to the block's name if available
```
Comment (marnovandermaas) [inline site/block-diagram/block_diagram.js:45]: Nit: maybe add periods at the end of comments in the JS file?

File: util/site/blocks.json [hunk 5]
```diff
@@ -227,8 +227,8 @@
         "href": "/hw/top_earlgrey/ip/ast#top",
         "report": null
     },
-    "padding": {
-        "name": "Padding",
+    "padring": {
+        "name": "Padring",
```
Comment (marnovandermaas) [inline util/site/blocks.json:231]: Can this commit be squashed so that this "padring" is not introduced in the commit history?

**Review by marnovandermaas — APPROVED** [review]
Just a few small comments. I mainly approve this PR because it  works well on the staging page. The only issue that I know of is that the horizontal arrows have a slightly different behavior on Chromium, but I think that can be fixed in a follow-up PR:
image

File: site/block-diagram/block_diagram.css (hunk 3 above)
Comment (jwnrt) [reply site/block-diagram/block_diagram.css:78]: Wow, good spot

**Comment by jwnrt** [comment]
Force pushed to fix the arrows and @marnovandermaas's comments.

The arrow problem came from Firefox giving `0` for unset variables, while Blink gives some kind of NaN that breaks calculations. Had a quick skim for other instances of that, but didn't see any obvious ones.

**Review by HU90m — APPROVED** [review]
Hahaha the old uninitialised variable.

**Review by hcallahan-lowrisc — APPROVED** [review]
> Force pushed to fix the arrows and @marnovandermaas's comments.
> 
> The arrow problem came from Firefox giving `0` for unset variables, while Blink gives some kind of NaN that breaks calculations. Had a quick skim for other instances of that, but didn't see any obvious ones.

Tested and LGTM! Thanks for all your work on this.

**Review by andreaskurth — APPROVED** [review]
LGTM based on what is currently visible on https://staging.opentitan.org/documentation/index.html. I didn't go through the code in detail as feedback iterations were already completed resulting in three approving reviews.

Very nice work overall, thanks @jwnrt!

I assume the Node-based diagram and its dependencies will be removed in a separate PR?

**Review by engdoreis — APPROVED** [review]

File: site/block-diagram/block_diagram.js [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        // Assign data to each block.
+        for (block of blocks) {
+            const tooltip = buildTooltip(stats[block.id]);
+            block.appendChild(tooltip);
+        }
+    };
+    req.open("GET", statsURL, true);
+    req.send();
+}
+/*
```
[the remark this answers is not in the document]
Comment (jwnrt) [reply site/block-diagram/block_diagram.js:26]: Oops, this looks like a merge conflict artefact. I'm going to improve this file's error handling in a future PR, so I'll clean up then.

**Pull request merged**